File: src/main/java/com/github/dkoval/leetcode/challenge/AddDigits.java
package com.github.dkoval.leetcode.challenge;

/**
 * <a href="https://leetcode.com/problems/add-digits/">Add Digits</a>
 * <p>
 * Given an integer num, repeatedly add all its digits until the result has only one digit, and return it.
 * <p>
 * Constraints:
 * <ul>
 *  <li>0 <= num <= 2^31 - 1</li>
 * </ul>
 * Follow up: Could you do it without any loop/recursion in O(1) runtime?
 */
public interface AddDigits {

    int addDigits(int num);

    class AddDigitsNaive implements AddDigits {

        @Override
        public int addDigits(int num) {
            int x = sumOfDigits(num);
            while (x > 9) {
                x = sumOfDigits(x);
            }
            return x;
        }

        private int sumOfDigits(int num) {
            int sum = 0;
            while (num > 0) {
                sum += num % 10;
                num /= 10;
            }
            return sum;
        }
    }
}

File: src/main/kotlin/com/github/dkoval/leetcode/challenge/AddDigits.kt
package com.github.dkoval.leetcode.challenge

object AddDigitsConstantTime : AddDigits {

    // Resources:
    // https://en.wikipedia.org/wiki/Digital_root
    // https://www.youtube.com/watch?v=tIjdI-ioXh0
    override fun addDigits(num: Int): Int = when {
        num == 0 -> 0
        num % 9 == 0 -> 9
        else -> num % 9
    }
}